### ISSUE TITLE
DNS records are not being created when outputting to Terraform

### ISSUE THREAD

**Issue by chrissnell** [kubernetes/kops#4834]
Thanks for submitting an issue! Please fill in as much of the template below as
you can.

------------- BUG REPORT TEMPLATE --------------------

**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

Version 1.9.0-alpha.3 (git-ad210dc4b)

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

1.9.3

**3. What cloud provider are you using?**

AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

```
kops create cluster     \
     --ssh-public-key=~/.ssh/my_ssh_key.pub \
     --authorization RBAC   \
     --node-count 3   \
     --zones "us-east-1a,us-east-1b,us-east-1c" \
     --master-zones "us-east-1a,us-east-1b,us-east-1c" \
     --node-size t2.large   \
     --master-size t2.medium  \
     --topology public  \
     --network-cidr=10.25.0.0/16 \
     --networking canal \
     --name us-east-1.kube.redacted.com
```

Observe the following line in the output log:

```
0328 16:12:15.805750   29693 dnszone.go:249] Existing zone "kube.redacted.com." found; will configure TF to reuse
```
OK, so far, so good.

Now run:

```kops update cluster us-east-1.kube.redacted.com--out tmpdir --target terraform```

Observe that no Terraform `aws_route53_record` resources are generated in the output.

Run `kops validate cluster us-east-1.kube.redacted.com` and watch it fail because a DNS record for `api.us-east-1.kube.redacted.com` does not exist.

Note: it makes no difference if you provide a `--dns-zone` flag to `kops create cluster`.  Either way, the resources are not created.  Obviously, if you let kops create the cluster directly (no Terraform), everything works fine.

**5. What happened after the commands executed?**

No Terraform `aws_route53_record` resources are created.

6. What did you expect to happen?

`aws_route53_record` resources would be created for the API and nodes.

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -oyaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```
apiVersion: kops/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: 2018-03-28T21:10:46Z
  name: us-east-1.kube.redacted.com
spec:
  api:
    dns: {}
  authorization:
    rbac: {}
  channel: stable
  cloudProvider: aws
  configBase: s3://redacted-kops-state-store/us-east-1.kube.redacted.com
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-us-east-1a
      name: a
    - instanceGroup: master-us-east-1b
      name: b
    - instanceGroup: master-us-east-1c
      name: c
    name: main
  - etcdMembers:
    - instanceGroup: master-us-east-1a
      name: a
    - instanceGroup: master-us-east-1b
      name: b
    - instanceGroup: master-us-east-1c
      name: c
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.9.3
  masterPublicName: api.us-east-1.kube.redacted.com
  networkCIDR: 10.25.0.0/16
  networking:
    canal: {}
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 0.0.0.0/0
  subnets:
  - cidr: 10.25.32.0/19
    name: us-east-1a
    type: Public
    zone: us-east-1a
  - cidr: 10.25.64.0/19
    name: us-east-1b
    type: Public
    zone: us-east-1b
  - cidr: 10.25.96.0/19
    name: us-east-1c
    type: Public
    zone: us-east-1c
  topology:
    dns:
      type: Public
    masters: public
    nodes: public

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-03-28T21:10:46Z
  labels:
    kops.k8s.io/cluster: us-east-1.kube.redacted.com
  name: master-us-east-1a
spec:
  image: kope.io/k8s-1.8-debian-jessie-amd64-hvm-ebs-2018-02-08
  machineType: t2.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-east-1a
  role: Master
  subnets:
  - us-east-1a

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-03-28T21:10:46Z
  labels:
    kops.k8s.io/cluster: us-east-1.kube.redacted.com
  name: master-us-east-1b
spec:
  image: kope.io/k8s-1.8-debian-jessie-amd64-hvm-ebs-2018-02-08
  machineType: t2.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-east-1b
  role: Master
  subnets:
  - us-east-1b

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-03-28T21:10:47Z
  labels:
    kops.k8s.io/cluster: us-east-1.kube.redacted.com
  name: master-us-east-1c
spec:
  image: kope.io/k8s-1.8-debian-jessie-amd64-hvm-ebs-2018-02-08
  machineType: t2.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-east-1c
  role: Master
  subnets:
  - us-east-1c

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-03-28T21:10:47Z
  labels:
    kops.k8s.io/cluster: us-east-1.kube.redacted.com
  name: nodes
spec:
  image: kope.io/k8s-1.8-debian-jessie-amd64-hvm-ebs-2018-02-08
  machineType: t2.large
  maxSize: 3
  minSize: 3
  nodeLabels:
    kops.k8s.io/instancegroup: nodes
  role: Node
  subnets:
  - us-east-1a
  - us-east-1b
  - us-east-1c
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

Gist available privately upon request.  

**9. Anything else do we need to know?**

**Comment by chrislovecnm** [comment]
DNS records are set by protokube and dns controller, which are components of kops.  This is so that during upgrades we can have the cluster manage its own dns.

https://github.com/kubernetes/kops/issues/3612 - provides more details, on how kops starts.

1. Look at the protokube logs, a container running on the master, and see if it is setting etcd dns correctly
2. If etcd and the API server is starting check that the dns-controller deployment is starting inside of the kube-system namespace.  I have seen resource issues where it could not start.  Rarely there are problems with your CNI provider.

Protokube sets the etcd dns entries, and dns-controller sets the api server dns entries.

**Comment by chrissnell** [comment]
OK, thanks for those tips.   I was able to SSH into the master and get a little bit of an idea about what's going on here.

For some reason, the dns-controller is not successfully starting until almost five minutes after etcd, kube-apiserver, kubelet, and the rest of the gang come up.  I can't figure out why.  I noticed that kubelet seems to be unable to talk to kube-apiserver at first and, in turn, kube-apiserver is unable to talk to etcd.   Everything eventually comes up.  

Here are some logs.  Hoping you can spot the problem here!

| service |  gist |
---------- | ---------------
| `protokube` | https://gist.github.com/chrissnell/d2ae88c48a74f50b5c67dba01a0708f1 |
|  `dnscontroller`  |   https://gist.github.com/chrissnell/39c00d03304a22bf122ea4a62934d400 |
| `kube-apiserver` |  https://gist.github.com/chrissnell/09d31b536bb355750d9e4a5f694a1879 |
| `etcd`  | https://gist.github.com/chrissnell/675120ebc56732e45d284f37133b7f5c |
| `nodeup`  |  https://gist.github.com/chrissnell/065b9b441214585733cf97aa26bf97fc |
| `kubelet`  | https://gist.github.com/chrissnell/02ce9ff25ed656e52d105478065ffcca  |

**Comment by chrissnell** [comment]
This may or may not be relevant:  when I build the cluster with CoreOS on all IGs, the cluster never comes up properly--not even after 10+ minutes--because etcd is never even started.

This is also something that worked when I built the cluster with `kops update` instead of `terraform apply`.

**Comment by chrissnell** [comment]
Digging further, my hunch is that etcd is part of the problem here.

Observations:

etcd is flapping on startup.  It appears to be having difficulty doing a DNS lookup on its own name record.  This is `etcd-b`, failing to resolve `etcd-b.internal.us-east-1-prod.kube.redacted.com`:  

https://gist.github.com/chrissnell/675120ebc56732e45d284f37133b7f5c#file-etcd-log-L17

This is odd, because the record was created by protokube almost 10 seconds prior:

https://gist.github.com/chrissnell/d2ae88c48a74f50b5c67dba01a0708f1#file-protokube-log-L95

Oddly, protokube continues to create that DNS record as the minutes tick on:

https://gist.github.com/chrissnell/d2ae88c48a74f50b5c67dba01a0708f1#file-protokube-log-L530

**Comment by chrissnell** [comment]
So, I figured out that the cluster **eventually** comes up...the API DNS record just takes much longer to come up when building from Terraform versus `kops update cluster`.  This, combined with my local DNS resolver caching NXDOMAIN records made it look like the API DNS was never coming up...but it was....slowly.

Closing for now.  If this concerns the authors, I'm happy to work with you to debug why it takes so long for dns-controller to populate the API Route53 record on a Terraform build.
